batch/batch/driver/pool.py

```diff
@@ -0,0 +1,469 @@
+                        await asyncio.gather(*[self.create_instance() for _ in range(instances_needed)])
+
+                n_live_instances = self.n_instances_by_state['pending'] + self.n_instances_by_state['active']
+                if (ENABLE_STANDING_WORKER
+                        and self.standing_worker
+                        and n_live_instances == 0
+                        and self.max_instances > 0):
+                    await self.create_instance(cores=self.standing_worker_cores,
+                                               max_idle_time_msecs=STANDING_WORKER_MAX_IDLE_TIME_MSECS)
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
+            except Exception:
+                log.exception('in control loop')
+            await asyncio.sleep(15)
+
+    def __str__(self):
+        return f'pool {self.name}'
+
+
+class ExceededSharesCounter:
+    def __init__(self):
+        self._global_counter = WindowFractionCounter(10)
+
+    def push(self, success: bool):
+        self._global_counter.push('exceeded_shares', success)
+
+    def rate(self) -> float:
+        return self._global_counter.fraction()
+
+    def __repr__(self):
+        return f'global {self._global_counter}'
+
+
+class PoolScheduler:
+    def __init__(self, app, pool):
+        self.app = app
+        self.scheduler_state_changed = pool.scheduler_state_changed
+        self.db = app['db']
+        self.pool = pool
+        self.async_worker_pool = AsyncWorkerPool(parallelism=100, queue_size=100)
+        self.exceeded_shares_counter = ExceededSharesCounter()
+        self.task_manager = aiotools.BackgroundTaskManager()
+
+    async def async_init(self):
+        self.task_manager.ensure_future(retry_long_running(
+            'schedule_loop',
+            run_if_changed, self.scheduler_state_changed, self.schedule_loop_body))
+
+    def shutdown(self):
+        try:
+            self.task_manager.shutdown()
+        finally:
+            self.async_worker_pool.shutdown()
+
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([
+            worker.free_cores_mcpu
+            for worker in self.pool.healthy_instances_by_free_cores
+        ])
+
+        user_running_cores_mcpu = {}
+        user_total_cores_mcpu = {}
+        result = {}
+
+        pending_users_by_running_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_running_cores_mcpu[user])
+        allocating_users_by_total_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_total_cores_mcpu[user])
+
+        records = self.db.execute_and_fetchall(
+            '''
+SELECT user,
+  CAST(COALESCE(SUM(n_ready_jobs), 0) AS SIGNED) AS n_ready_jobs,
+  CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu,
+  CAST(COALESCE(SUM(n_running_jobs), 0) AS SIGNED) AS n_running_jobs,
+  CAST(COALESCE(SUM(running_cores_mcpu), 0) AS SIGNED) AS running_cores_mcpu
```

Comment:
I'm pretty sure these are all NOT NULL.

Reply:
See comment above.

Reply:
You're indeed right about the behavior above, which sends me into fits of rage, but here the `GROUP BY` ensures there are no empty sets of rows so these values always have a well-defined value.
```
mysql> drop temporary table if exists foo  ; create temporary table foo (user varchar(100) not null, n_ready_jobs int(11) not null , primary key(user));
Query OK, 0 rows affected, 1 warning (0.00 sec)

Query OK, 0 rows affected (0.00 sec)

mysql> select sum(n_ready_jobs) from foo;
+-------------------+
| sum(n_ready_jobs) |
+-------------------+
|              NULL |
+-------------------+
1 row in set (0.01 sec)

mysql> select sum(n_ready_jobs) from foo group by user;
Empty set (0.01 sec)
```

Reply:
But I guess we can leave it if the uniformity seems good. I'm in such a fit of rage right now.